### PR TITLE
[git-webkit] Cannot create pull request branch using bug URL, fails to fetch comments from Bugzilla, fails to modify bug

### PR DESCRIPTION
**Pull request by mcatanzaro**
#### b2a78a3c5258302b4e6e4e978e5b38e940f0587d
<pre>
[git-webkit] Cannot create pull request branch using bug URL, fails to fetch comments from Bugzilla, fails to modify bug
<a href="https://bugs.webkit.org/show_bug.cgi?id=239741">https://bugs.webkit.org/show_bug.cgi?id=239741</a>

Patch by Michael Catanzaro &lt;mcatanzaro@gnome.org &gt; on 2022-04-26
Reviewed by Jonathan Bedard.

* Tools/Scripts/libraries/webkitbugspy/setup.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.credentials.validater):
(Tracker._login_arguments):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla.__init__):
(Bugzilla.request):

Canonical link: <a href="https://commits.webkit.org/250028@main">https://commits.webkit.org/250028@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@293495">https://svn.webkit.org/repository/webkit/trunk@293495</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py [hunk 1]
```diff
@@ -112,7 +112,9 @@ def from_string(self, string):
 
     def credentials(self, required=True, validate=False):
         def validater(username, password):
-            response = requests.get('{}/rest/user/{}?login={}&password={}'.format(self.url, username, username, password))
+            quoted_username = requests.utils.quote(username)
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: `test-webkitpy` tests are failing. Need to change `CREDENTIAL_RE` in `Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py` too.

**Review by JonWBedard — COMMENTED** [review]
Could we also bump the minor version of the library in `Tools/Scripts/libraries/webkitbugspy/setup.py` and `Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py` too?

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py (hunk 1 above)
Comment (mcatanzaro) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: Oh no, not a regex :(

**Comment by mcatanzaro** [comment]
BTW another improvement would be for git-webkit to issue fewer requests when one fails (ideally just stop after the first failed request), to avoid getting IP banned from WebKit Bugzilla, but I will leave that for someone more familiar with this code. ;)

**Comment by mcatanzaro** [comment]
> Could we also bump the minor version of the library in `Tools/Scripts/libraries/webkitbugspy/setup.py` and `Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py` too?

Sure.

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py (hunk 1 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: It's just used for testing, so if you pass the unit tests, you're good!

Comment (mcatanzaro) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: OK I've entered the test-webkitpy rabbit hole.

I will deal with https://bugs.webkit.org/show_bug.cgi?id=239786.

Maybe you could look at https://bugs.webkit.org/show_bug.cgi?id=239788?

Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: I will look at that one...that's probably just laziness when writing the unit test on my part.

Comment (mcatanzaro) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: > Need to change CREDENTIAL_RE in Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py too.

Are you sure something is wrong with that regex? I see a bunch of tests are failing, but it doesn't look related to the regex.

Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: If you look at https://ews-build.webkit.org/#/builders/5/builds/78706, something like `webkitbugspy.tests.bugzilla_unittest.TestBugzilla.test_create` is definitely failing because you've changed what the requests to bugzilla look like.

That's all `CREDENTIAL_RE` is doing, by the way, we're matching a string in the request in our mock server to extract the username and password.

Comment (mcatanzaro) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: It was actually failing _before_ checking `CREDENTIAL_RE`.

Anyway, you suggested a better solution for resolving this -- URL decode in the mock implementation -- and that worked.

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py [hunk 2]
```diff
@@ -54,7 +55,7 @@ def __init__(self, hostname='bugs.example.com', users=None, issues=None, environ
 
         prefix = self.hosts[0].replace('.', '_').upper()
         self._environment = environment or mocks.Environment(**{
-            '{}_USERNAME'.format(prefix): 'usernmae',
+            '{}_USERNAME'.format(prefix): 'username',
```
Comment (mcatanzaro) [inline Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py]: Note: I added a drive-by fix here. I assume this is not doing anything, because it doesn't seem to affect test results.

**Review by JonWBedard — APPROVED** [review]
